File: app/src/androidTest/java/com/kira/kdownloader/util/DownloadDirectoryScannerAndroidTest.java
package com.kira.kdownloader.util;

import android.content.ContentUris;
import android.content.ContentValues;
import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.MediaStore;

import androidx.room.Room;
import androidx.test.core.app.ApplicationProvider;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.kira.kdownloader.data.AppDatabase;
import com.kira.kdownloader.data.DownloadEntity;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.OutputStream;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assume.assumeTrue;

@RunWith(AndroidJUnit4.class)
public class DownloadDirectoryScannerAndroidTest {
    @Test public void mediaStoreFileInDefaultDirectoryIsImportedOnce() throws Exception {
        assumeTrue(Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q);
        Context context = ApplicationProvider.getApplicationContext();
        AppDatabase database = Room.inMemoryDatabaseBuilder(context, AppDatabase.class).build();
        Uri mediaUri = null;
        try {
            String displayName = "history-scan-" + System.currentTimeMillis() + ".mp4";
            ContentValues values = new ContentValues();
            values.put(MediaStore.MediaColumns.DISPLAY_NAME, displayName);
            values.put(MediaStore.MediaColumns.MIME_TYPE, "video/mp4");
            values.put(MediaStore.MediaColumns.RELATIVE_PATH, Environment.DIRECTORY_DOWNLOADS + "/KDownloads");
            values.put(MediaStore.MediaColumns.IS_PENDING, 1);
            mediaUri = context.getContentResolver().insert(MediaStore.Downloads.EXTERNAL_CONTENT_URI, values);
            assertNotNull(mediaUri);
            try (OutputStream output = context.getContentResolver().openOutputStream(mediaUri)) {
                assertNotNull(output);
                output.write(new byte[]{0, 0, 0, 20, 102, 116, 121, 112});
            }
            ContentValues complete = new ContentValues();
            complete.put(MediaStore.MediaColumns.IS_PENDING, 0);
            context.getContentResolver().update(mediaUri, complete, null, null);

            int firstImportCount = DownloadDirectoryScanner.syncIntoHistory(context, database.downloadDao(), true);
            long insertedMediaId = ContentUris.parseId(mediaUri);
            DownloadEntity imported = findByMediaId(database, insertedMediaId);
            assertNotNull(imported);
            assertEquals("VIDEO", imported.getKind());
            assertEquals(0, DownloadDirectoryScanner.syncIntoHistory(context, database.downloadDao(), true));
            assertEquals(imported, findByMediaId(database, insertedMediaId));
            if (firstImportCount < 1) throw new AssertionError("Expected at least one imported row");
        } finally {
            if (mediaUri != null) context.getContentResolver().delete(mediaUri, null, null);
            database.close();
        }
    }

    private DownloadEntity findByMediaId(AppDatabase database, long mediaId) {
        for (DownloadEntity row : database.downloadDao().getAllSync()) {
            try {
                if (row.getFileUri() != null && ContentUris.parseId(Uri.parse(row.getFileUri())) == mediaId) return row;
            } catch (RuntimeException ignored) {}
        }
        return null;
    }
}

File: app/src/androidTest/java/com/kira/kdownloader/util/DownloadDirectoryScannerAndroidTest.kt
package com.kira.kdownloader.util

import android.content.ContentValues
import android.content.ContentUris
import android.content.Context
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.kira.kdownloader.data.AppDatabase
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assume.assumeTrue
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class DownloadDirectoryScannerAndroidTest {
    @Test
    fun mediaStoreFileInDefaultDirectoryIsImportedOnce() = runBlocking {
        assumeTrue(Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)

        val context = ApplicationProvider.getApplicationContext<Context>()
        val database = Room.inMemoryDatabaseBuilder(context, AppDatabase::class.java).build()
        var mediaUri: Uri? = null

        try {
            val displayName = "history-scan-${System.currentTimeMillis()}.mp4"
            mediaUri = context.contentResolver.insert(
                MediaStore.Downloads.EXTERNAL_CONTENT_URI,
                ContentValues().apply {
                    put(MediaStore.MediaColumns.DISPLAY_NAME, displayName)
                    put(MediaStore.MediaColumns.MIME_TYPE, "video/mp4")
                    put(
                        MediaStore.MediaColumns.RELATIVE_PATH,
                        "${Environment.DIRECTORY_DOWNLOADS}/KDownloads",
                    )
                    put(MediaStore.MediaColumns.IS_PENDING, 1)
                },
            )
            assertNotNull(mediaUri)

            context.contentResolver.openOutputStream(requireNotNull(mediaUri)).use { output ->
                requireNotNull(output).write(byteArrayOf(0, 0, 0, 20, 102, 116, 121, 112))
            }
            context.contentResolver.update(
                requireNotNull(mediaUri),
                ContentValues().apply { put(MediaStore.MediaColumns.IS_PENDING, 0) },
                null,
                null,
            )

            val firstImportCount = DownloadDirectoryScanner.syncIntoHistory(
                context,
                database.downloadDao(),
                force = true,
            )
            val insertedMediaId = ContentUris.parseId(requireNotNull(mediaUri))
            val imported = database.downloadDao().observeAll().first()
                .singleOrNull { row ->
                    runCatching {
                        ContentUris.parseId(Uri.parse(row.fileUri)) == insertedMediaId
                    }.getOrDefault(false)
                }

            assertNotNull(imported)
            assertEquals("VIDEO", imported?.kind)
            assertEquals(
                0,
                DownloadDirectoryScanner.syncIntoHistory(
                    context,
                    database.downloadDao(),
                    force = true,
                ),
            )
            assertEquals(imported, database.downloadDao().observeAll().first()
                .singleOrNull { row ->
                    runCatching {
                        ContentUris.parseId(Uri.parse(row.fileUri)) == insertedMediaId
                    }.getOrDefault(false)
                })
            check(firstImportCount >= 1)
        } finally {
            mediaUri?.let { uri -> context.contentResolver.delete(uri, null, null) }
            database.close()
        }
    }
}
